{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting blast_seq.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile blast_seq.py\n",
    "\n",
    "\"\"\"\n",
    "The goal for this program is to provide an iterable function that can be run\n",
    "across a reference genome to isolate and return sequences that lie between two\n",
    "primer sequences provided, given certain variables.\n",
    "The function takes the following files:\n",
    "\n",
    "    The reference genome path string \n",
    "        Either as a fasta or fastq file ending in that file name\n",
    "\n",
    "    The primers file that contains both the forward and reverse primer sequences\n",
    "\n",
    "    Parameters for determining whether the region selected is ITS or TEF\n",
    "        Implemented as a maximum and minimum length of the extracted sequence\n",
    "            This allows for extraction of other regions\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "import subprocess\n",
    "import os\n",
    "import argparse\n",
    "import fileinput\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "The goal for this program is to provide an iterable function that can be run\n",
    "across a reference genome to isolate and return sequences that lie between two\n",
    "primer sequences provided, given certain variables.\n",
    "The function takes the following files:\n",
    "\n",
    "    The reference genome path string \n",
    "        Either as a fasta or fastq file ending in that file name\n",
    "\n",
    "    The primers file that contains both the forward and reverse primer sequences\n",
    "\n",
    "    Parameters for determining whether the region selected is ITS or TEF\n",
    "        Implemented as a maximum and minimum length of the extracted sequence\n",
    "            This allows for extraction of other regions\n",
    "\n",
    "\"\"\")\n",
    "parser.add_argument(\"reference_genome\", help=\"The reference genome to scan for the particular sequence\")\n",
    "parser.add_argument(\"sequence\", help=\"The sequence from elsewhere to match\")\n",
    "args = parser.parse_args()\n",
    "reference = args.reference_genome\n",
    "print(reference[32:-16])\n",
    "species_name = reference[32:-16]\n",
    "\n",
    "database = reference[32:-16]+\"/\"+reference[32:-6]+'_db'\n",
    "cmd = 'makeblastdb -in %s -dbtype nucl -out %s' % (reference, database)\n",
    "subprocess.getoutput(cmd)\n",
    "\n",
    "outfmt6 = reference[32:-16]+\"/\"+reference[32:-6]+'.outfmt6'\n",
    "cmd2 = 'blastn -query %s -db %s -evalue=100000 -task \"blastn\" -outfmt 6 > %s' % (args.sequence, database, outfmt6)\n",
    "subprocess.getoutput(cmd2)\n",
    "\n",
    "extract = pd.read_csv(outfmt6, sep=\"\\t\", header=None, names=[\"qseqid\", \"sseqid\", \"pident\", \"length\", \"mismatch\", \"gapopen\", \"qstart\", \"qend\", \"sstart\", \"send\", \"evalue\", \"bitscore\"])\n",
    "\n",
    "extract_df = extract.loc[extract['evalue']==min(extract['evalue'])]\n",
    "extract_df = extract_df.reset_index()\n",
    "\n",
    "extract_bed = pd.DataFrame(columns=['chrom', 'chromStart', 'chromEnd'])\n",
    "for i in range(0, len(extract_df)):\n",
    "    if extract_df['sstart'][i] <= extract_df['send'][i]:\n",
    "        extract_bed = extract_bed.append({'chrom': extract_df['sseqid'][i],'chromStart': (extract_df['sstart'][i])-1, 'chromEnd': extract_df['send'][i]}, ignore_index=True)\n",
    "    else:\n",
    "        extract_bed = extract_bed.append({'chrom': extract_df['sseqid'][i],'chromStart': (extract_df['send'][i])-1, 'chromEnd': extract_df['sstart'][i]}, ignore_index=True)\n",
    "extract_bed.sort_values(['chromStart'])\n",
    "\n",
    "\n",
    "bedfile = reference[32:-16]+\"/\"+reference[32:-6]+'.ITS.bedfile'\n",
    "\n",
    "extract_bed.to_csv(bedfile, sep='\\t', header=False, index=False)\n",
    "\n",
    "bedoutput = reference[32:-16]+\"/\"+reference[32:-6]+'.ITS.bedoutput.fasta'\n",
    "cmd3 = 'bedtools getfasta -fo %s -fi %s -bed %s' % (bedoutput, reference, bedfile)\n",
    "subprocess.getoutput(cmd3)\n",
    "\n",
    "f = open(bedoutput, 'r')\n",
    "filedata = f.read()\n",
    "f.close()\n",
    "new_data = filedata.replace('>', '>%s:' % species_name)\n",
    "f = open(bedoutput, 'w')\n",
    "f.write(new_data)\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
